File: src/lib/oauth.ts
import {
  CLAUDE_OAUTH_CLIENT_ID,
  CLAUDE_OAUTH_AUTHORIZE_URL,
  CLAUDE_OAUTH_TOKEN_URL,
  CLAUDE_OAUTH_REDIRECT_URI,
  CLAUDE_OAUTH_SCOPES,
  DEFAULT_CALLBACK_PORT,
} from './constants'
import { generatePKCE, type PKCEPair } from './pkce'
import { saveCredentials, loadCredentials, type ClaudeOAuthCredentials } from './storage'

interface TokenResponse {
  access_token: string
  refresh_token: string
  expires_in: number
  token_type: string
}

let pendingPKCE: PKCEPair | null = null

export interface OAuthFlowResult {
  authUrl: string
  codeVerifier: string
}

export function startOAuthFlow(): OAuthFlowResult {
  const pkce = generatePKCE()
  pendingPKCE = pkce

  const url = new URL(CLAUDE_OAUTH_AUTHORIZE_URL)
  url.searchParams.set('code', 'true')
  url.searchParams.set('client_id', CLAUDE_OAUTH_CLIENT_ID)
  url.searchParams.set('response_type', 'code')
  url.searchParams.set('redirect_uri', CLAUDE_OAUTH_REDIRECT_URI)
  url.searchParams.set('scope', CLAUDE_OAUTH_SCOPES)
  url.searchParams.set('code_challenge', pkce.challenge)
  url.searchParams.set('code_challenge_method', 'S256')
  url.searchParams.set('state', pkce.verifier)

  return {
    authUrl: url.toString(),
    codeVerifier: pkce.verifier,
  }
}

export async function exchangeCodeForTokens(
  authorizationCode: string,
  codeVerifier?: string
): Promise<ClaudeOAuthCredentials> {
  const verifier = codeVerifier ?? pendingPKCE?.verifier
  if (!verifier) {
    throw new Error('No code verifier found. Start OAuth flow first.')
  }

  const [code, state] = authorizationCode.trim().split('#')

  const response = await fetch(CLAUDE_OAUTH_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      code,
      state,
      grant_type: 'authorization_code',
      client_id: CLAUDE_OAUTH_CLIENT_ID,
      redirect_uri: CLAUDE_OAUTH_REDIRECT_URI,
      code_verifier: verifier,
    }),
  })

  if (!response.ok) {
    const errorText = await response.text()
    throw new Error(`Token exchange failed: ${errorText}`)
  }

  const data = (await response.json()) as TokenResponse
  pendingPKCE = null

  const credentials: ClaudeOAuthCredentials = {
    accessToken: data.access_token,
    refreshToken: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    connectedAt: Date.now(),
  }

  await saveCredentials(credentials)
  return credentials
}

export async function refreshAccessToken(
  credentials: ClaudeOAuthCredentials
): Promise<ClaudeOAuthCredentials> {
  const response = await fetch(CLAUDE_OAUTH_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      grant_type: 'refresh_token',
      refresh_token: credentials.refreshToken,
      client_id: CLAUDE_OAUTH_CLIENT_ID,
    }),
  })

  if (!response.ok) {
    const errorText = await response.text()
    throw new Error(`Token refresh failed: ${errorText}`)
  }

  const data = (await response.json()) as TokenResponse

  const newCredentials: ClaudeOAuthCredentials = {
    accessToken: data.access_token,
    refreshToken: data.refresh_token ?? credentials.refreshToken,
    expiresAt: Date.now() + data.expires_in * 1000,
    connectedAt: credentials.connectedAt,
  }

  await saveCredentials(newCredentials)
  return newCredentials
}

export async function getValidAccessToken(): Promise<string> {
  let credentials = await loadCredentials()
  if (!credentials) {
    throw new Error('Not authenticated. Run login first.')
  }

  const bufferMs = 5 * 60 * 1000
  if (credentials.expiresAt <= Date.now() + bufferMs) {
    credentials = await refreshAccessToken(credentials)
  }

  return credentials.accessToken
}

export async function startLocalCallbackServer(
  port: number = DEFAULT_CALLBACK_PORT
): Promise<string> {
  return new Promise((resolve, reject) => {
    const server = Bun.serve({
      port,
      fetch(req) {
        const url = new URL(req.url)
        if (url.pathname === '/callback') {
          const code = url.searchParams.get('code')
          const error = url.searchParams.get('error')

          if (error) {
            reject(new Error(`OAuth error: ${error}`))
            setTimeout(() => server.stop(), 100)
            return new Response('Authentication failed. You can close this window.', {
              headers: { 'Content-Type': 'text/html' },
            })
          }

          if (code) {
            resolve(code)
            setTimeout(() => server.stop(), 100)
            return new Response(
              '<html><body><h1>Success!</h1><p>You can close this window.</p></body></html>',
              { headers: { 'Content-Type': 'text/html' } }
            )
          }
        }

        return new Response('Not found', { status: 404 })
      },
    })

    setTimeout(() => {
      server.stop()
      reject(new Error('OAuth callback timeout (5 minutes)'))
    }, 5 * 60 * 1000)
  })
}

File: python/claude_oauth/oauth.py
"""OAuth flow handling for Claude API authentication."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import TYPE_CHECKING
from urllib.parse import urlencode

import httpx

from .constants import (
    CLAUDE_OAUTH_AUTHORIZE_URL,
    CLAUDE_OAUTH_CLIENT_ID,
    CLAUDE_OAUTH_REDIRECT_URI,
    CLAUDE_OAUTH_SCOPES,
    CLAUDE_OAUTH_TOKEN_URL,
)
from .pkce import PKCEPair, generate_pkce
from .storage import ClaudeOAuthCredentials, load_credentials, save_credentials

if TYPE_CHECKING:
    pass


@dataclass
class OAuthFlowResult:
    auth_url: str
    code_verifier: str


_pending_pkce: PKCEPair | None = None


def start_oauth_flow() -> OAuthFlowResult:
    global _pending_pkce
    pkce = generate_pkce()
    _pending_pkce = pkce

    params = {
        "code": "true",
        "client_id": CLAUDE_OAUTH_CLIENT_ID,
        "response_type": "code",
        "redirect_uri": CLAUDE_OAUTH_REDIRECT_URI,
        "scope": CLAUDE_OAUTH_SCOPES,
        "code_challenge": pkce.challenge,
        "code_challenge_method": "S256",
        "state": pkce.verifier,
    }
    auth_url = f"{CLAUDE_OAUTH_AUTHORIZE_URL}?{urlencode(params)}"
    return OAuthFlowResult(auth_url=auth_url, code_verifier=pkce.verifier)


def exchange_code_for_tokens(
    authorization_code: str, code_verifier: str | None = None
) -> ClaudeOAuthCredentials:
    global _pending_pkce

    verifier = code_verifier or (_pending_pkce.verifier if _pending_pkce else None)
    if not verifier:
        raise ValueError("No code verifier found. Start OAuth flow first.")

    code, state = authorization_code.strip().split("#", 1) if "#" in authorization_code else (authorization_code.strip(), "")

    with httpx.Client() as client:
        response = client.post(
            CLAUDE_OAUTH_TOKEN_URL,
            json={
                "code": code,
                "state": state,
                "grant_type": "authorization_code",
                "client_id": CLAUDE_OAUTH_CLIENT_ID,
                "redirect_uri": CLAUDE_OAUTH_REDIRECT_URI,
                "code_verifier": verifier,
            },
        )

    if not response.is_success:
        raise RuntimeError(f"Token exchange failed: {response.text}")

    data = response.json()
    _pending_pkce = None

    credentials = ClaudeOAuthCredentials(
        access_token=data["access_token"],
        refresh_token=data["refresh_token"],
        expires_at=_current_time_ms() + data["expires_in"] * 1000,
        connected_at=_current_time_ms(),
    )

    save_credentials(credentials)
    return credentials


def refresh_access_token(credentials: ClaudeOAuthCredentials) -> ClaudeOAuthCredentials:
    with httpx.Client() as client:
        response = client.post(
            CLAUDE_OAUTH_TOKEN_URL,
            json={
                "grant_type": "refresh_token",
                "refresh_token": credentials.refresh_token,
                "client_id": CLAUDE_OAUTH_CLIENT_ID,
            },
        )

    if not response.is_success:
        raise RuntimeError(f"Token refresh failed: {response.text}")

    data = response.json()

    new_credentials = ClaudeOAuthCredentials(
        access_token=data["access_token"],
        refresh_token=data.get("refresh_token", credentials.refresh_token),
        expires_at=_current_time_ms() + data["expires_in"] * 1000,
        connected_at=credentials.connected_at,
    )

    save_credentials(new_credentials)
    return new_credentials


def get_valid_access_token() -> str:
    credentials = load_credentials()
    if not credentials:
        raise RuntimeError("Not authenticated. Run login first.")

    buffer_ms = 5 * 60 * 1000
    if credentials.expires_at <= _current_time_ms() + buffer_ms:
        credentials = refresh_access_token(credentials)

    return credentials.access_token


def _current_time_ms() -> int:
    return int(time.time() * 1000)
